lib/salesforce_bulk_api/connection.rb

```diff
@@ -5,62 +5,87 @@ class Connection
     include Concerns::Throttling
 
     LOGIN_HOST = 'login.salesforce.com'
+    RESTFORCE_CLIENT_TYPE = 'Restforce::Data::Client'
 
     def initialize(api_version, client)
       @client = client
       @api_version = api_version
       @path_prefix = "/services/async/#{@api_version}/"
 
-      login()
+      login
     end
 
-    def login()
-      client_type = @client.class.to_s
-      case client_type
-      when "Restforce::Data::Client"
+    def login
+      @client_type = @client.class.to_s
+
+      case @client_type
+      when RESTFORCE_CLIENT_TYPE
+        # authenticate the client if it's not already authenticated
+        @client.authenticate! if @client.options[:oauth_token].empty? || @client.options[:instance_url].empty?
+
         @session_id = @client.options[:oauth_token]
         @server_url = @client.options[:instance_url]
       else
         @session_id = @client.oauth_token
         @server_url = @client.instance_url
       end
-      @instance = parse_instance()
+
+      @instance = parse_instance
       @instance_host = "#{@instance}.salesforce.com"
     end
 
+    def reauthenticate
+      # currently only supports Restforce reauthentication
+      @client.authenticate! if @client_type == RESTFORCE_CLIENT_TYPE
+      login
+    end
+
     def post_xml(host, path, xml, headers)
-      host = host || @instance_host
-      if host != LOGIN_HOST # Not login, need to add session id to header
-        headers['X-SFDC-Session'] = @session_id
-        path = "#{@path_prefix}#{path}"
-      end
-      i = 0
+      retried = false
       begin
+        host = host || @instance_host
+        if host != LOGIN_HOST # Not login, need to add session id to header
+          headers['X-SFDC-Session'] = @session_id
+          path = "#{@path_prefix}#{path}"
+        end
+
         count :post
         throttle(http_method: :post, path: path)
-        https(host).post(path, xml, headers).body
+        response = https(host).post(path, xml, headers)
+        raise if response.code == '401'
+
+        response.body
       rescue
-        i += 1
-        if i < 3
-          puts "Request fail #{i}: Retrying #{path}"
-          retry
-        else
-          puts "FATAL: Request to #{path} failed three times."
-          raise
-        end
```

Comment:
This essentially never did anything because `Net::HTTP`'s `post` method doesn't throw exceptions.